File: linux2/src/auth/invite-manager.ts
import crypto from 'node:crypto';

interface InviteEntry {
  code: string;
  createdAt: Date;
  expiresAt: Date;
}

const INVITE_TTL_MS = 5 * 60 * 1000; // 5 minutes
const CODE_LENGTH = 8;

export class InviteManager {
  private readonly invites = new Map<string, InviteEntry>();

  generateInviteCode(): { code: string; expiresAt: Date } {
    this.cleanupExpired();

    const bytes = crypto.randomBytes(CODE_LENGTH);
    const code = bytes.toString('base64url').slice(0, CODE_LENGTH);
    const now = new Date();
    const expiresAt = new Date(now.getTime() + INVITE_TTL_MS);

    this.invites.set(code, { code, createdAt: now, expiresAt });
    return { code, expiresAt };
  }

  /** Validate and consume a one-time invite code. Returns true if valid. */
  consumeInviteCode(code: string): boolean {
    this.cleanupExpired();
    const entry = this.invites.get(code);
    if (!entry) return false;
    if (new Date() > entry.expiresAt) {
      this.invites.delete(code);
      return false;
    }
    // One-time use: delete after consumption
    this.invites.delete(code);
    return true;
  }

  isValidInviteCode(code: string): boolean {
    const entry = this.invites.get(code);
    if (!entry) return false;
    return new Date() <= entry.expiresAt;
  }

  clearAll(): void {
    this.invites.clear();
  }

  private cleanupExpired(): void {
    const now = new Date();
    for (const [key, entry] of this.invites) {
      if (now > entry.expiresAt) {
        this.invites.delete(key);
      }
    }
  }
}

File: pc/src/PhoneShell.Core/Services/InviteManager.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;

namespace PhoneShell.Core.Services;

/// <summary>
/// Manages one-time invite codes for adding devices to a group.
/// Codes are 8-character base64url strings, valid for 5 minutes, single-use.
/// </summary>
public sealed class InviteManager
{
    private readonly ConcurrentDictionary<string, InviteEntry> _invites = new();
    private static readonly TimeSpan InviteTtl = TimeSpan.FromMinutes(5);
    private const int CodeLength = 8;

    public (string Code, DateTimeOffset ExpiresAt) GenerateInviteCode()
    {
        CleanupExpired();

        var bytes = RandomNumberGenerator.GetBytes(CodeLength);
        var code = Convert.ToBase64String(bytes)
            .Replace('+', '-')
            .Replace('/', '_')
            .TrimEnd('=')[..CodeLength];

        var now = DateTimeOffset.UtcNow;
        var expiresAt = now.Add(InviteTtl);

        _invites[code] = new InviteEntry(code, now, expiresAt);
        return (code, expiresAt);
    }

    /// <summary>
    /// Check if an invite code is valid without consuming it.
    /// </summary>
    public bool IsValidInviteCode(string code)
    {
        CleanupExpired();

        if (!_invites.TryGetValue(code, out var entry))
            return false;

        return DateTimeOffset.UtcNow <= entry.ExpiresAt;
    }

    /// <summary>
    /// Validate and consume a one-time invite code. Returns true if valid.
    /// </summary>
    public bool ConsumeInviteCode(string code)
    {
        CleanupExpired();

        if (!_invites.TryRemove(code, out var entry))
            return false;

        return DateTimeOffset.UtcNow <= entry.ExpiresAt;
    }

    public void ClearAll()
    {
        _invites.Clear();
    }

    private void CleanupExpired()
    {
        var now = DateTimeOffset.UtcNow;
        foreach (var kvp in _invites)
        {
            if (now > kvp.Value.ExpiresAt)
                _invites.TryRemove(kvp.Key, out _);
        }
    }

    private sealed record InviteEntry(string Code, DateTimeOffset CreatedAt, DateTimeOffset ExpiresAt);
}
